File: src/16.最接近的三数之和.py
#
# @lc app=leetcode.cn id=16 lang=python3
#
# [16] 最接近的三数之和
# two-pointers

from math import fabs
from typing import List

# @lc code=start
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums, n = sorted(nums), len(nums)
        minx, res = 1000000000, 0
        i = 0
        while i < n-2:
            j, k = i+1, i+2
            while k < n-1 and nums[i] + nums[j] + nums[k] < target:
                k += 1
            while j < k:
                tmp = nums[i] + nums[j] + nums[k]
                x = fabs(tmp-target)
                if x < minx:
                    minx, res = x, tmp
                if tmp > target:
                    k -= 1
                elif tmp < target:
                    j += 1
                elif tmp == target:
                    return target
            i += 1
        return res

# @lc code=end

s = Solution()
print(s.threeSumClosest([1,2,4,8,16,32,64,128], 82))
print(s.threeSumClosest([0,0,0], 1))
print(s.threeSumClosest([-1,2,1,-4], 1))

File: src/16.最接近的三数之和.go
package main

/*
 * @lc app=leetcode.cn id=16 lang=golang
 *
 * [16] 最接近的三数之和
 */
// two-pointers

import (
	"fmt"
	"math"
	"sort"
)

// @lc code=start
func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	n := len(nums)
	ans, minDiff := 0, math.MaxInt32
	for i := 0; i < n-2; i++ {
		for j, k := i+1, n-1; j < k; {
			if diff := target - (nums[i] + nums[j] + nums[k]); diff >= 0 {
				if diff < minDiff {
					ans, minDiff = nums[i]+nums[j]+nums[k], diff
				}
				j++
			} else {
				if -diff < minDiff {
					ans, minDiff = nums[i]+nums[j]+nums[k], -diff
				}
				k--
			}
		}
	}
	return ans
}

// @lc code=end
func main() {
	fmt.Println(threeSumClosest([]int{-1, 2, 1, -4}, 1))
}
